core/commands/pubsub.go

```diff
@@ -119,20 +96,51 @@ To use, the daemon must be run with '--enable-pubsub-experiment'.
 				connectToPubSubPeers(req.Context(), n, cid)
 			}()
 		}
+
+		for {
+			msg, err := sub.Next(req.Context())
+			if err == io.EOF || err == context.Canceled {
+				return
+			} else if err != nil {
+				re.SetError(err, cmdsutil.ErrNormal)
+				return
+			}
+
+			re.Emit(msg)
+		}
 	},
-	Marshalers: cmds.MarshalerMap{
-		cmds.Text: getPsMsgMarshaler(func(m *floodsub.Message) (io.Reader, error) {
-			return bytes.NewReader(m.Data), nil
+	Encoders: map[cmds.EncodingType]func(cmds.Request) func(io.Writer) cmds.Encoder{
+		cmds.Text: cmds.MakeEncoder(func(w io.Writer, v interface{}) error {
+			m, ok := v.(*floodsub.Message)
```

Comment:
Yeah I think using the `EncoderMap` and `EncoderFunc` would do the trick. When I said I still need some cleanup I meant this kind of stuff.